rpc/http.go

```diff
@@ -0,0 +1,268 @@
+
+	go func() {
+		if err := fasthttp.Serve(h.listener, reqHandler); err != nil {
+			h.Logger.Fatal().Msgf("error while serving http: %v", err)
+		}
+	}()
+	return nil
+}
+
+// mainHandler is the handler called by the fasthttp server when a proper http
+// request is received. It calls the appropriate handler based on transport type
+func (h *HttpServer) mainHandler(ctx *fasthttp.RequestCtx) {
+	r := &http.Request{}
+	if err := fasthttpadaptor.ConvertRequest(ctx, r, false); err != nil {
+		ctx.Error("Internal Server Error", fasthttp.StatusInternalServerError)
+		return
+	}
+
+	// HttpServer can handle both http and ws requests. If http and ws configurations use the same port, then single
+	// HttpServer is running and handling all requests. If they have different ports, then 2 separate servers are running.
+	// The following code differentiates the cases based on the configurations and Config.WsOnly variable
+	if strings.EqualFold(h.Config.HttpEndpoint, h.Config.WsEndpoint) {
+		if isWebsocket(r) {
+			h.fastWsHandler(ctx)
+		} else if !h.Config.WsOnly {
+			h.fastHTTPHandler(ctx, r)
+		} else {
+			ctx.Error("", fasthttp.StatusNotFound)
+		}
+	} else {
+		h.fastHTTPHandler(ctx, r)
+	}
+}
+
+// fastWSHandler is the handler to serve WS requests
+func (h *HttpServer) fastWsHandler(ctx *fasthttp.RequestCtx) {
+	err := upgrader.Upgrade(ctx, func(conn *websocket.Conn) {
+		defer conn.Close()
+		wsCtx := &WebSocketContext{ws: conn, subscriptions: make(map[ID]*Subscription), subscriptionsMtx: sync.Mutex{}}
+		wsCtx.output = make(chan []byte, 100)
+		wsCtx.closed = make(chan struct{})
+
+		wsCtx.outputWg.Add(1)
+		go h.handleWebSocketOutput(wsCtx)
+
+		for {
+			messageType, message, err := wsCtx.ws.ReadMessage()
+			if err != nil {
+				break
+			}
+			if messageType != websocket.TextMessage {
+				h.Logger.Warn().Msgf("websocket: got message with unknown type #%v", messageType)
+				continue
+			}
+			// get the clientIp and add it to context so rpcserver can use it when needed
+			clientIp := ctx.RemoteIP()
+			cCtx := utils.PutClientIpKey(ctx, clientIp)
+			resp := h.resolver.ResolveWs(&cCtx, wsCtx, message)
+			if resp != nil {
+				wsCtx.output <- resp
+			}
+		}
+
+		close(wsCtx.closed)
+		h.resolver.CloseWsConn(wsCtx)
+		close(wsCtx.output)
+		wsCtx.outputWg.Wait()
+	})
+
+	if err != nil {
+		if _, ok := err.(websocket.HandshakeError); ok {
+			h.Logger.Error().Err(err).Msg("error upgrading to websocket")
+		}
+		return
+	}
+}
+
+// handleWebSocketOutput listens the websocket output channel and writes incoming data to connection
+func (h *HttpServer) handleWebSocketOutput(wsCtx *WebSocketContext) {
+	defer wsCtx.outputWg.Done()
+	defer func() {
+		for range wsCtx.output {
+		}
+	}()
+
+	for {
+		select {
+		case <-wsCtx.closed:
+			return
+		default:
+		}
+
+		select {
+		case <-wsCtx.closed:
+			return
+		case data, ok := <-wsCtx.output:
+			if !ok {
+				return
+			}
+			wsCtx.ws.WriteMessage(websocket.TextMessage, data)
```

Comment:
I was also indecisive here. The generated error is not super helpful but can be at least logged. However, this part is taken from Terminator as is and I assumed it is not good to do any operation in `select` and decided not to touch since it is working.

Reply:
I think, better be logging